### ISSUE TITLE
AsDiscrete, Activations allow batched input as before

### ISSUE THREAD

**Issue by myron** [Project-MONAI/MONAI#3683]
It seems that as of MONAI 0.6 the  AsDiscrete, Activations don't take batched data, and now require to decollete before use, and then reassemble after use. This is too complicated for most applications and creates unnecessary code. 

https://github.com/Project-MONAI/MONAI/issues/2549
https://github.com/Project-MONAI/MONAI/commit/2085a49820be146ad953ece23041540a237e08b6
https://github.com/Project-MONAI/tutorials/blob/master/modules/decollate_batch.ipynb

Please add an optional parameter (dim=0) to AsDiscrete, Activations, which I can manually set to dim=1 when needed to operate on batched data directly. 

Thanks

**Comment by wyli** [comment]
Hi @myron could you elaborate on "reassemble after use"?

**Comment by myron** [comment]
after decollation, it run post-processing, and the output is a list of images. So in most cases I need to assemble (collate) back into a batched representation (single tensor) to compute the final loss or accuracy. Even all standard Pytorch loss functions and accuracy functions take a batch (not a list of tensors). So here deviate from Pytorch login. 

I'm not sure why we did this decollate at all (it's rare that we need different postprocessing for different images in the same batch).

**Comment by Nic-Ma** [comment]
Hi @myron ,

One of the common usage of decollate is that: we can decollate the predictions then **invert** preprocessing transforms for every item as the preprocessing **random transforms** applied on every input image may be different and the **input image shape** may also be different. Then we can apply other processings for every prediction separately, like `AsDiscrete`, etc. An example is here:
https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/torch/unet_inference_dict.py#L70
Thanks.

**Comment by myron** [comment]
Hi, so I appreciate the flexibility that such approach provides.  But honestly the use case when it's needed are very small. What applications do we need to invert the random transforms? AsDiscrete (provides ability to do softmax or one_hot), when do we ever need it to be applied separately or differently for each prediction?  If such use cases exist,  they are very limited compared to 99% of cases of standard segmentation/classification, where the "decollate" creates unnecessary code (and has additional compute overhead).  If you insist on having decollate, at least add the optional "dim " parameter to AsDiscrete, Activations, so that I can use them directly without decollate. thank you

**Comment by wyli** [comment]
Sure, let's add an optional `dim=0`. 
The consistent data format would reduce input dimensionality validation code and make it easy to build robust workflows. Also, perhaps in your use case, you want to use `torch.argmax` or `monai.networks.one_hot` directly... (In the end `AsDiscrete` `Actions` are very thin wrappers of the underlying pytorch API). And for the same reason you can use or ignore the decollate mechanism.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Currently, all the MONAI transforms are for channel-first data instead of batch data, for this special use case, maybe you can easily achieve it by `Lamda` transform?

Thanks.

**Comment by Nic-Ma** [comment]
Hi @myron ,

1. Could you please help share with us your full program to help understand why `decollate` is not needed?
I feel for most postprocessing cases of segmentation, we `Activation`, `Invert spatial transforms`, `AsDiscrete`, `SaveImage`, `MeanDice`, etc.
2. I think it's better to invert the spatial transforms(spacing, resize, etc.) back to the original input shape to compute MeanDice with the original labels. @dongyang0122 Could you please help confirm it?
3. If you are using MONAI SupervisedEvaluator or Clara configs, the `decollate` is default logic in it, you don't need to write the code:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/engines/workflow.py#L204
4. And you can use `Lambda` transform to call `torch.softmax()` function directly.

Thanks.

**Comment by dongyang0122** [comment]
It is possible that different neural networks are trained under different imaging spacing for specific applications. It is better to revert the spacing of results for fair comparison.

**Comment by myron** [comment]
So, for many applications data is already resampled (e.g brats) so there is nothing to resample or invert.  And even if data is being resampled on the fly  - during Training epoch (I don't need to resample/invert anything back to compute metric, I'd rather improve speed and compute it at e.g.1x1x1 resolution), same for Validation epoch.  I suppose, at the very end, during inference if we want labels to be at original resolution then we need to invert resampling. 

May be we can make the decollate to be a Transform ?  Can I define my post-tranforms as Compose ([Decolate, Activation])?

**Comment by myron** [comment]
it seems other people are confused about it too https://github.com/Project-MONAI/MONAI/discussions/4374, why these functions don't work with batched data
